drivers: flash: flash_mspi_nor: support runtime flash probe

danieldegrasse opened the pull request:
Support runtime probe of flash data within the MSPI NOR flash driver.

Since SFDP probe data is often not available for 1-4-4 or 1-8-8 mode, the approach used here is to instead define a table of flash configurations, with the JEDEC ID as the key.

We can also choose to conditionally compile each of these flash definitions in, using Kconfigs (to reduce flash size)

swift-tk commented:
I still don’t think it is the best approach to categorize flash_cmd by jedec ID. RAM/ROM usage can get uncontrollably large as more device is added. Not to mention that lots of cmds are the same. 

Instead, I think we should group by flash vendors because in general user would know where they bought the chip from but not specific part number. So we would let user to choose a vendor by kconfig and then enable this runtime probing by jedec ID which would largely reduce unnecessary memory usage.

The most efficient way is certainly SFDP, but since it would only work for absolutely jedec compliance devices, the runtime probing may run parallel to SFDP to provide quirks.

swift-tk commented:
We should probably invite broader review and possibly discuss this in architecture review meetings.

danieldegrasse commented:
> I still don’t think it is the best approach to categorize flash_cmd by jedec ID. RAM/ROM usage can get uncontrollably large as more device is added. Not to mention that lots of cmds are the same.
> 
> Instead, I think we should group by flash vendors because in general user would know where they bought the chip from but not specific part number. So we would let user to choose a vendor by kconfig and then enable this runtime probing by jedec ID which would largely reduce unnecessary memory usage.
> 
> The most efficient way is certainly SFDP, but since it would only work for absolutely jedec compliance devices, the runtime probing may run parallel to SFDP to provide quirks.

I agree the ROM table can get quite large. Maybe we could create per-device KConfigs to enable each device in the ROM table? Then users would have two methods of enabling support for their flash chip:
- assign correct compatible for their flash chip in devicetree, then only the Kconfig for their chip is on, and the ROM table has one entry
- turn on a Kconfig like `CONFIG_MSPI_NOR_ALL_DEVICES` that enables all devices for runtime probing

I think the ROM table can be a good system provided we do something like the above- If we make the size of the table only one entry based on the devicetree compatible, than we can probably stop calling this runtime probing and instead use the ROM table for supporting a variety of devices.


> We should probably invite broader review and possibly discuss this in architecture review meetings.

Fine to do that if you'd like, do you want to tag this with the `arch-review` label?

swift-tk commented:
> I agree the ROM table can get quite large. Maybe we could create per-device KConfigs to enable each device in the ROM table? Then users would have two methods of enabling support for their flash chip:
> 
> * assign correct compatible for their flash chip in devicetree, then only the Kconfig for their chip is on, and the ROM table has one entry
> * turn on a Kconfig like `CONFIG_MSPI_NOR_ALL_DEVICES` that enables all devices for runtime probing

Then wouldn't we end up with a whole bunch of if defines in the code if creating kconfig per device?

> I think the ROM table can be a good system provided we do something like the above- If we make the size of the table only one entry based on the devicetree compatible, than we can probably stop calling this runtime probing and instead use the ROM table for supporting a variety of devices.

I agree, however, having different cmds for the same flash operations is not the only thing that differs. The ROM table can not take care of difference in terms of operational sequence.  For example:

- some flash need to check flag status after reset and has the option to also check for it after write or erase along with the regular status register checking
- some flash need to send extra command when clocking it at higher speed.
- some flash have extra protect bit for each block that need to set/clear after write/read

Having different cmds for the standard jedec specific operation also fall into quirks, so I think that the ROM table should only exist in those device that needed it and queried from a quirks API as the devices that are jedec compliant 100% in terms of cmds would never need it.

> Fine to do that if you'd like, do you want to tag this with the `arch-review` label?

Yes, sure. In the mean time, we should invte more people to participate the discussion.

danieldegrasse commented:
> > I agree the ROM table can get quite large. Maybe we could create per-device KConfigs to enable each device in the ROM table? Then users would have two methods of enabling support for their flash chip:
> > 
> > * assign correct compatible for their flash chip in devicetree, then only the Kconfig for their chip is on, and the ROM table has one entry
> > * turn on a Kconfig like `CONFIG_MSPI_NOR_ALL_DEVICES` that enables all devices for runtime probing
> 
> Then wouldn't we end up with a whole bunch of if defines in the code if creating kconfig per device?

Yes, my vision would be something like this:

```
struct flash_mspi_nor_devs mspi_nor_devs[] = {
#ifdef CONFIG_MSPI_NOR_MT35XU02GCBA
	{
		.jedec_id = {0x2C, 0x5B, 0x1A}, /* MT35XU02GCBA */
		.page_size = 4096,
		.flash_size = 0x4000000,
		.dev_cfg = {
			.io_mode = MSPI_IO_MODE_OCTAL_1_8_8,
			.data_rate = MSPI_DATA_RATE_SINGLE,
			.endian = MSPI_XFER_BIG_ENDIAN,
		},
    ...
   },
#endif
#ifdef CONFIG_MSPI_NOR_MT25QU512ABB
	{
		.jedec_id = {0x20, 0xBB, 0x20}, /* MT25QU512ABB */
		.page_size = 4096,
		.flash_size = 0x4000000,
		.dev_cfg = {
			.io_mode = MSPI_IO_MODE_QUAD_1_4_4,
			.data_rate = MSPI_DATA_RATE_SINGLE,
			.endian = MSPI_XFER_BIG_ENDIAN,
		},
     ...
     },
#endif
};
```

Personally I think this is a pretty clean approach to supporting multiple flash devices in one driver, but I'm open to other suggestions.

> 
> > I think the ROM table can be a good system provided we do something like the above- If we make the size of the table only one entry based on the devicetree compatible, than we can probably stop calling this runtime probing and instead use the ROM table for supporting a variety of devices.
> 
> I agree, however, having different cmds for the same flash operations is not the only thing that differs. The ROM table can not take care of difference in terms of operational sequence. For example:
> 
>     * some flash need to check flag status after reset and has the option to also check for it after write or erase along with the regular status register checking
> 
>     * some flash need to send extra command when clocking it at higher speed.
> 
>     * some flash have extra protect bit for each block that need to set/clear after write/read
> 

Sure- I think we need to determine some design that lets drivers with quirks replace some API functions, while reusing common code.

> 
> Having different cmds for the standard jedec specific operation also fall into quirks, so I think that the ROM table should only exist in those device that needed it and queried from a quirks API as the devices that are jedec compliant 100% in terms of cmds would never need it.

I disagree here. The issue is that a full SFDP probe for compliant devices will add a lot of overhead, which a single entry ROM table won't.

danieldegrasse commented:
What about a design like this:

- we implement the read/write/erase functions for the MSPI NOR driver to use a ROM table for command sets
- at boot time, if there is one entry in the ROM table (IE others are compiled out), we use that
- otherwise we probe the JEDEC ID, and if an entry is found in the ROM table, we use that
- if no JEDEC ID is found (and CONFIG_MSPI_NOR_RUNTIME_PROBE is on), we use SFDP parameters to configure the flash

For flash devices with quirks we can't describe in the ROM table, we expect those devices to implement custom MSPI nor drivers- those drivers can reuse the read/write/erase functions from the base MSPI NOR driver or replace them if needed

swift-tk commented:
> Having different cmds for the standard jedec specific operation also fall into quirks, so I think that the ROM table should only exist in those device that needed it and queried from a quirks API as the devices that are jedec compliant 100% in terms of cmds would never need it.
> 
> I disagree here. The issue is that a full SFDP probe for compliant devices will add a lot of overhead, which a single entry ROM table won't.

I realize that for jedec compliant devices, the ROM table is more efficient than SFDP probe. However, this ROM table would be the same for all these devices and would unnecessarily increase ROM usage with `CONFIG_MSPI_NOR_ALL_DEVICES` if the same ROM table is repeatly defined under different device kconfig.

> I think we need to determine some design that lets drivers with quirks replace some API functions, while reusing common code.
> For flash devices with quirks we can't describe in the ROM table, we expect those devices to implement custom MSPI nor drivers- those drivers can reuse the read/write/erase functions from the base MSPI NOR driver or replace them if needed

I was thinking about adding quirk APIs like pre_write, post_write, pre_read, post_read, post_rst and call them pointers in the generic nor flash driver when they are defined so we don't need the custom drivers.

- pre_write : Get the write cmd, dummy and etc, and conduct any additional operations required from vendor. 
   - unprotect by address
   - possible frequency/mode switching and special high speed requirement
- post_write: Conduct any additional operations other than checking the flash status register.
    - possible frequency/mode switching
- pre_read: Get the read cmd, dummy. Possible frequency/mode switching
- post_read: Possible frequency/mode switching
- post_rst: wait delay and additional status checking, getting/initializing the cmd struct
- ......

Underneath the quirk post_rst, when getting the cmd struct, we could then query by jedec ID, frequency, io_mode and etc. 
The quirk may include operation in the optional DW of the jedec spec which is provided by SFDP or explicitly written in the vendor.c

The only thing that I'm concerned with is it can get pretty messy in one core driver.

danieldegrasse commented:
> > Having different cmds for the standard jedec specific operation also fall into quirks, so I think that the ROM table should only exist in those device that needed it and queried from a quirks API as the devices that are jedec compliant 100% in terms of cmds would never need it.
> > I disagree here. The issue is that a full SFDP probe for compliant devices will add a lot of overhead, which a single entry ROM table won't.
> 
> I realize that for jedec compliant devices, the ROM table is more efficient than SFDP probe. However, this ROM table would be the same for all these devices and would unnecessarily increase ROM usage with `CONFIG_MSPI_NOR_ALL_DEVICES` if the same ROM table is repeatly defined under different device kconfig.
> 

This is a great point. Maybe the way we do this is define "base tables", that devices can point to? So something like the following:

```
const struct flash_cmd_table 1_1_1_jedec_compliant = {
	.read_cmd = {.cmd = 0x3, ..},
	.write_cmd = {.cmd = 0x12, ..},
};

const struct flash_cmd_table 1_4_4_jedec_compliant = {
	.read_cmd = {..},
	.write_cmd = {..},
};

const struct flash_cmd_table mx35 = {
	.read_cmd = {..},
	.write_cmd = {..},
};

const struct flash_jedec_table rom_table[] = {
	IF_ENABLED(CONFIG_FLASH_MSPI_NOR_MT25xxx, {.jedec_id = 0x205403, .flash_table = &1_4_4_jedec_compliant}),
	IF_ENABLED(CONFIG_FLASH_MSPI_NOR_IS25 ,{.jedec_id = 0x204300 .flash_table = &1_1_1_jedec_compliant}),
	IF_ENABLED(CONFIG_FLASH_MSPI_NOR_MX35xxx ,{.jedec_id = 0x403300 .flash_table = &mx35}),
};
```

This way we don't duplicate things for flash devices that are shared- the compiler should optimize out unused "base tables"


> > I think we need to determine some design that lets drivers with quirks replace some API functions, while reusing common code.
> > For flash devices with quirks we can't describe in the ROM table, we expect those devices to implement custom MSPI nor drivers- those drivers can reuse the read/write/erase functions from the base MSPI NOR driver or replace them if needed
> 
> I was thinking about adding quirk APIs like pre_write, post_write, pre_read, post_read, post_rst and call them pointers in the generic nor flash driver when they are defined so we don't need the custom drivers.
> 
>     * pre_write : Get the write cmd, dummy and etc, and conduct any additional operations required from vendor.
>       
>       * unprotect by address
>       * possible frequency/mode switching and special high speed requirement
> 
>     * post_write: Conduct any additional operations other than checking the flash status register.
>       
>       * possible frequency/mode switching
> 
>     * pre_read: Get the read cmd, dummy. Possible frequency/mode switching
> 
>     * post_read: Possible frequency/mode switching
> 
>     * post_rst: wait delay and additional status checking, getting/initializing the cmd struct
> 
>     * ......
> 
> 
> Underneath the quirk post_rst, when getting the cmd struct, we could then query by jedec ID, frequency, io_mode and etc. The quirk may include operation in the optional DW of the jedec spec which is provided by SFDP or explicitly written in the vendor.c
> 
> The only thing that I'm concerned with is it can get pretty messy in one core driver.

Sure, this could work. My concern is that some flash devices even get to the point where they don't use the same flash status register. I think it makes more sense to expose the majority of the mspi_nor driver functions in a header, so that other drivers can use them. IE if we split up the write function like so:

```
int flash_mspi_nor_write() {
	for (i = 0; i < num_blocks; i++) {
		flash_mspi_nor_write_block();
		flash_mspi_nor_wait_busy();
	}
}
```
We'd expose all these functions in a header to make it possible to reuse them. Thoughts on this?

swift-tk commented:
> Sure, this could work. My concern is that some flash devices even get to the point where they don't use the same flash status register. I think it makes more sense to expose the majority of the mspi_nor driver functions in a header, so that other drivers can use them. IE if we split up the write function like so:
> ```
> int flash_mspi_nor_write() {
> 	for (i = 0; i < num_blocks; i++) {
> 		flash_mspi_nor_write_block();
> 		flash_mspi_nor_wait_busy();
> 	}
> }
> ```
> We'd expose all these functions in a header to make it possible to reuse them. Thoughts on this?

That would make the jedec core a utility instead of a driver? I don’t know if we can have it both ways. Additionally, what would be the point to centralized the cmd distribution by ROM table if we are going to setup device specific drivers anyways… We could just pass the transfer struct with everything already prepared(cmd, dummy and etc.) to the common parts.

danieldegrasse commented:
> > Sure, this could work. My concern is that some flash devices even get to the point where they don't use the same flash status register. I think it makes more sense to expose the majority of the mspi_nor driver functions in a header, so that other drivers can use them. IE if we split up the write function like so:
> > ```
> > int flash_mspi_nor_write() {
> > 	for (i = 0; i < num_blocks; i++) {
> > 		flash_mspi_nor_write_block();
> > 		flash_mspi_nor_wait_busy();
> > 	}
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > We'd expose all these functions in a header to make it possible to reuse them. Thoughts on this?
> 
> That would make the jedec core a utility instead of a driver? I don’t know if we can have it both ways.

I think we can have it both ways- make the JEDEC core a utility, and write a driver that uses those utility functions. Other drivers can then use the functions as they see fit.

>  Additionally, what would be the point to centralized the cmd distribution by ROM table if we are going to setup device specific drivers anyways… We could just pass the transfer struct with everything already prepared(cmd, dummy and etc.) to the common parts.

The value from my POV is supporting runtime probe (IE a case where you want to be able to support multiple flash devices with one binary image, which is currently possible in the SPI NOR driver using SFDP runtime probing)

swift-tk commented:
> I think we can have it both ways- make the JEDEC core a utility, and write a driver that uses those utility functions. Other drivers can then use the functions as they see fit.

That is exactly my point, the jedec core is no longer a driver but can still work. However, in the case where I want to support multiple flash device in the same image, then several drivers will be compiled altogether instead of one? 

> The value from my POV is supporting runtime probe (IE a case where you want to be able to support multiple flash devices with one binary image, which is currently possible in the SPI NOR driver using SFDP runtime probing)

Where would we draw boundaries then?  To what extend, could we call these devices as a group that can be supported in one image? It seems to me that having device specific drivers goes againt this intent.

> My concern is that some flash devices even get to the point where they don't use the same flash status register

To counter this, we could further abstract the jedec core as APIs and in their `post_init quirks` API, the `flash_mspi_nor_wait_busy` can be replaced/overloaded then.

In summary, I think a single uniformed driver with quirks API and jedec API calls could extend the maximum compatibility.

danieldegrasse commented:
> > I think we can have it both ways- make the JEDEC core a utility, and write a driver that uses those utility functions. Other drivers can then use the functions as they see fit.
> 
> That is exactly my point, the jedec core is no longer a driver but can still work. However, in the case where I want to support multiple flash device in the same image, then several drivers will be compiled altogether instead of one?
> 
> > The value from my POV is supporting runtime probe (IE a case where you want to be able to support multiple flash devices with one binary image, which is currently possible in the SPI NOR driver using SFDP runtime probing)
> 
> Where would we draw boundaries then? To what extend, could we call these devices as a group that can be supported in one image? It seems to me that having device specific drivers goes againt this intent.
> 
> > My concern is that some flash devices even get to the point where they don't use the same flash status register
> 
> To counter this, we could further abstract the jedec core as APIs and in their `post_init quirks` API, the `flash_mspi_nor_wait_busy` can be replaced/overloaded then. SFDP may also be able to handle some difference, but not all. So it there will probably be separae SFDP utility where the cmd, bit position and etc can be altered before calling the jedec core APIs.
> 
> In summary, I think a single uniformed driver with quirks API and jedec API calls could extend the maximum compatibility.

Alright, if you'd prefer to add a quirks API to the uniformed driver I'm alright with that approach. My initial thought was that we would have the following:
- JEDEC core APIs
- Generic driver that reuses JEDEC core APIs, and pulls flash command sets from a ROM table
- Device specific drivers for any flash chip that requires more specialized operations than can be supported within a ROM table

If have a quirks API, it sounds like we'd need the following (based on the list you provided earlier):
- pre_write : Get the write cmd, dummy and etc, and conduct any additional operations required from vendor
- post_write: Conduct any additional operations other than checking the flash status register.
- pre_read: Get the read cmd, dummy. Possible frequency/mode switching
- post_read: Possible frequency/mode switching
- post_rst: wait delay and additional status checking, getting/initializing the cmd struct

To move this PR forwards, I can do the following:
- add these quirk APIs into the driver
- move to using "base tables" for commands that flash drivers can reuse if they have the same command set but different JEDEC IDs
- add support for conditional compilation of each entry in the ROM table

Does that seem reasonable?

swift-tk commented:
> > > I think we can have it both ways- make the JEDEC core a utility, and write a driver that uses those utility functions. Other drivers can then use the functions as they see fit.
> > 
> > 
> > That is exactly my point, the jedec core is no longer a driver but can still work. However, in the case where I want to support multiple flash device in the same image, then several drivers will be compiled altogether instead of one?
> > > The value from my POV is supporting runtime probe (IE a case where you want to be able to support multiple flash devices with one binary image, which is currently possible in the SPI NOR driver using SFDP runtime probing)
> > 
> > 
> > Where would we draw boundaries then? To what extend, could we call these devices as a group that can be supported in one image? It seems to me that having device specific drivers goes againt this intent.
> > > My concern is that some flash devices even get to the point where they don't use the same flash status register
> > 
> > 
> > To counter this, we could further abstract the jedec core as APIs and in their `post_init quirks` API, the `flash_mspi_nor_wait_busy` can be replaced/overloaded then. SFDP may also be able to handle some difference, but not all. So it there will probably be separae SFDP utility where the cmd, bit position and etc can be altered before calling the jedec core APIs.
> > In summary, I think a single uniformed driver with quirks API and jedec API calls could extend the maximum compatibility.
> 
> Alright, if you'd prefer to add a quirks API to the uniformed driver I'm alright with that approach. My initial thought was that we would have the following:
> 
> * JEDEC core APIs
> * Generic driver that reuses JEDEC core APIs, and pulls flash command sets from a ROM table
> * Device specific drivers for any flash chip that requires more specialized operations than can be supported within a ROM table
> 
> If we have a quirks API, it sounds like we'd need the following (based on the list you provided earlier):
> 
> * pre_write : Get the write cmd, dummy and etc, and conduct any additional operations required from vendor
> * post_write: Conduct any additional operations other than checking the flash status register.
> * pre_read: Get the read cmd, dummy. Possible frequency/mode switching
> * post_read: Possible frequency/mode switching
> * post_rst: wait delay and additional status checking, getting/initializing the cmd struct
> 
> To move this PR forwards, I can do the following:
> 
> * add these quirk APIs into the driver
> * move to using "base tables" for commands that flash drivers can reuse if they have the same command set but different JEDEC IDs
> * add support for conditional compilation of each entry in the ROM table
> 
> Does that seem reasonable?

An additional pre_init maybe useful to override any jedec API.

Where would you like to put the quirk APIs implementation? My suggestion was to catalog them by vendors.

The rest is great. Thank you!

de-nordic reviewed (commented):
I do not mind having runtime auto configuration of driver, but I also have several issues here:
- the table will get huge, as more devices get supported, both in size and code in a single file
- jedec id may get something like 17 bytes long, with a lot of continuation bytes

I think that that the discovery should be split into units that cover only specific vendor and included by kconfigs.
Each vendor should implement own jedec id discovery function that will account for the actual vendor id that they have, which may not always follow the jedec id specification.

Ideally each vendor specific unit would be called in some order to check if what is on spi is something that can be recognized; once vendor specific code can confirm that it recognizes what it sees, then it inits dev data and breaks the loop.

So basically query and id check should be done in a vendor specific manner rather than in generic support code.

It vendor discovery source code should get own Kconfig for inclusion into compilation.

swift-tk commented:
> I do not mind having runtime auto configuration of driver, but I also have several issues here:
> 
> * the table will get huge, as more devices get supported, both in size and code in a single file
> * jedec id may get something like 17 bytes long, with a lot of continuation bytes
> 
> I think that that the discovery should be split into units that cover only specific vendor and included by kconfigs. Each vendor should implement own jedec id discovery function that will account for the actual vendor id that they have, which may not always follow the jedec id specification.
> 
> Ideally each vendor specific unit would be called in some order to check if what is on spi is something that can be recognized; once vendor specific code can confirm that it recognizes what it sees, then it inits dev data and breaks the loop.
> 
> So basically query and id check should be done in a vendor specific manner rather than in generic support code.
> 
> It vendor discovery source code should get own Kconfig for inclusion into compilation.

Then it makes even more sense to catalogue by vendor.  The actual discovery function from each vendor could be registered into an array of function pointers by the kconfigs(same way with ROM table) and looped in a jedec API. But the ID may have to be read multiple times because of different length requirement. For the ROM table, there would be one per vendor and may not be strictly grouped by jedec ID.

danieldegrasse committed:
drivers: flash_mspi_nor: move flash driver to subdirectory

Since mspi_nor driver consists of several files, move these files
into a subdirectory to keep the driver folder cleaner.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse committed:
drivers: flash: flash_mspi_nor: store flash data in struct

Store all flash data that could change at runtime in a struct. This will
permit us to store this data in a mutable section when runtime probing
is enabled.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse committed:
drivers: flash_mspi_nor: support runtime probe

Support runtime probe of flash devices in flash_mspi_nor driver. This is
enabled by defining tables of vendor devices that can be enabled via
opt-in Kconfigs, and will allow the driver to probe multiple flash
devices by JEDEC ID at runtime. If runtime probe is disabled, a specific
flash device can still be selected via the compatible string at compile
time.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

swift-tk commented on drivers/flash/flash_mspi_nor/flash_mspi_nor_micron.c:
```diff
@@ -0,0 +1,372 @@
+	&mxicy_mx25u_single,
+	&mxicy_mx25u_octal,
+	&mxicy_mx25r_single,
+	&mxicy_mx25r_quad,
```
Is it possible to have one entry here per device instead of per device and io mode?

swift-tk reviewed (commented):
Has the change been tested yet on any platform?
@masz-nordic Could you try this and see if it still works?

danieldegrasse commented:
> Has the change been tested yet on any platform?

Not upstream, no- this change has been tested on a custom ASIC, supported by this repo: https://github.com/tenstorrent/tt-zephyr-platforms. The ASIC uses the SPI DW controller, so effectively the change has been tested with that driver.

swift-tk reviewed (commented):
Note that the `commands_octal` in flash_mspi_nor_generic is not so generic as it is tailored towards mx25 series by @anangl. 
Noticable non-generic setting would be the dummy cycle in SFDP which 20 but won't be true for other flash devices like ATXP032.

danieldegrasse commented:
> Note that the commands_octal in flash_mspi_nor_generic is not so generic as it is tailored towards mx25 series by @anangl.
Noticable non-generic setting would be the dummy cycle in SFDP which 20 but won't be true for other flash devices like ATXP032.

True- I'll make the `commands_octal` a custom definition for the mx25.

danieldegrasse commented:
> True- I'll make the commands_octal a custom definition for the mx25.

This has been applied in latest push. I have also moved flash data that is not mode-specific into a shared struct for each flash device

swift-tk reviewed (dismissed):
LGTM

danieldegrasse commented:
Rebased
